data-prepper-plugins/kinesis-source/src/main/java/org/opensearch/dataprepper/plugins/source/kinesis/processor/KinesisRecordProcessor.java

```diff
@@ -0,0 +1,180 @@
+        final PluginSetting codecPluginSettings = new PluginSetting(codecConfiguration.getPluginName(), codecConfiguration.getPluginSettings());
+        this.codec = pluginFactory.loadPlugin(InputCodec.class, codecPluginSettings);
+        this.acknowledgementSetManager = acknowledgementSetManager;
+        this.acknowledgementSetCallbackCounter = pluginMetrics.counter(ACKNOWLEDGEMENT_SET_CALLBACK_METRIC_NAME);
+        this.recordProcessingErrors = pluginMetrics.counterWithTags(KINESIS_RECORD_PROCESSING_ERRORS, KINESIS_STREAM_TAG_KEY, streamIdentifier.streamName());
+        this.checkpointFailures = pluginMetrics.counterWithTags(KINESIS_CHECKPOINT_FAILURES, KINESIS_STREAM_TAG_KEY, streamIdentifier.streamName());
+        this.checkpointIntervalMilliSeconds = kinesisStreamConfig.getCheckPointIntervalInMilliseconds();
+        this.buffer = buffer;
+    }
+
+    private KinesisStreamConfig getStreamConfig(final KinesisSourceConfig kinesisSourceConfig) {
+        return kinesisSourceConfig.getStreams().stream().filter(streamConfig -> streamConfig.getName().equals(streamIdentifier.streamName())).findAny().get();
+    }
+
+    @Override
+    public void initialize(InitializationInput initializationInput) {
+        // Called once when the processor is initialized.
+        kinesisShardId = initializationInput.shardId();
+        LOG.info("Initialize Processor for shard: " + kinesisShardId);
+        lastCheckpointTimeInMillis = System.currentTimeMillis();
+    }
+
+    private AcknowledgementSet createAcknowledgmentSet(final ProcessRecordsInput processRecordsInput) {
+        return acknowledgementSetManager.create((result) -> {
+            acknowledgementSetCallbackCounter.increment();
+            if (result) {
+                LOG.info("acknowledgements received");
+                checkpoint(processRecordsInput.checkpointer());
+            } else {
+                LOG.info("acknowledgements received with false");
+            }
+
+        }, ACKNOWLEDGEMENT_SET_TIMEOUT);
+    }
+
+    @Override
+    public void processRecords(ProcessRecordsInput processRecordsInput) {
+        List<Record<Event>> records = new ArrayList<>();
+
+        try {
+            AcknowledgementSet acknowledgementSet;
+            boolean acknowledgementsEnabled = kinesisSourceConfig.isAcknowledgments();
+            if (acknowledgementsEnabled) {
+                acknowledgementSet = createAcknowledgmentSet(processRecordsInput);
+            } else {
+                acknowledgementSet = null;
+            }
+
+            for (KinesisClientRecord record : processRecordsInput.records()) {
+                processRecord(record, records::add);
+            }
+
+            if (acknowledgementSet != null) {
+                records.forEach(record -> {
+                    acknowledgementSet.add(record.getData());
+                });
+            }
+
+            buffer.writeAll(records, bufferTimeoutMillis);
```

Comment:
Why are we writing type KinesisClientRecord to the buffer? Shouldn't this be of type `Event`? I don't see the location where events are currently constructed, including where you would add the metadata keys, external origination time, and tags.

Reply:
Please refer to this (https://github.com/opensearch-project/data-prepper/pull/4836/files#diff-c2a354679cb457978f2a208ae1f342e03d44b283298c72d41e35db59d728c96bR167-R176) in the latest change. It should have the metadata set. I have not included the origination time which I will do in a follow-up change.